go/libraries/doltcore/sqle/dprocedures/dolt_rebase.go

```diff
@@ -0,0 +1,539 @@
+	rebaseState := workingSet.RebaseState()
+	workingSet = workingSet.AbortRebase()
+	err = doltSession.SetWorkingSet(ctx, ctx.GetCurrentDatabase(), workingSet)
+	if err != nil {
+		return err
+	}
+
+	// Delete the working branch
+	var rsc doltdb.ReplicationStatusController
+	dbData, ok := doltSession.GetDbData(ctx, ctx.GetCurrentDatabase())
+	if !ok {
+		return fmt.Errorf("unable to get DbData for database %s", ctx.GetCurrentDatabase())
+	}
+	headRef, err := doltSession.CWBHeadRef(ctx, ctx.GetCurrentDatabase())
+	if err != nil {
+		return err
+	}
+	err = actions.DeleteBranch(ctx, dbData, headRef.GetPath(), actions.DeleteOptions{
+		Force:                      true,
+		AllowDeletingCurrentBranch: true,
+	}, doltSession.Provider(), &rsc)
+	if err != nil {
+		return err
+	}
+
+	// Switch back to the original branch head
+	wsRef, err := ref.WorkingSetRefForHead(ref.NewBranchRef(rebaseState.Branch()))
+	if err != nil {
+		return err
+	}
+
+	return doltSession.SwitchWorkingSet(ctx, ctx.GetCurrentDatabase(), wsRef)
+}
+
+func continueRebase(ctx *sql.Context) error {
+	// TODO: Eventually, when we allow interactive-rebases to be stopped and started (e.g. with the break action,
+	//       or for conflict resolution), we'll need to track what step we're at in the rebase plan.
+
+	// Validate that we are in an interactive rebase
+	doltSession := dsess.DSessFromSess(ctx.Session)
+	workingSet, err := doltSession.WorkingSet(ctx, ctx.GetCurrentDatabase())
+	if err != nil {
+		return err
+	}
+	if !workingSet.RebaseActive() {
+		return fmt.Errorf("no rebase in progress")
+	}
+
+	db, err := doltSession.Provider().Database(ctx, ctx.GetCurrentDatabase())
+	if err != nil {
+		return err
+	}
+
+	rdb, ok := db.(rebase.RebasePlanDatabase)
+	if !ok {
+		return fmt.Errorf("expected a dsess.RebasePlanDatabase implementation, but received a %T", db)
+	}
+	rebasePlan, err := rdb.LoadRebasePlan(ctx)
+	if err != nil {
+		return err
+	}
+
+	err = rebase.ValidateRebasePlan(ctx, rebasePlan)
+	if err != nil {
+		return err
+	}
+
+	for _, step := range rebasePlan.Steps {
+		err = processRebasePlanStep(ctx, &step)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Update the branch being rebased to point to the same commit as our temporary working branch
+	rebaseBranchWorkingSet, err := doltSession.WorkingSet(ctx, ctx.GetCurrentDatabase())
+	if err != nil {
+		return err
+	}
+	dbData, ok := doltSession.GetDbData(ctx, ctx.GetCurrentDatabase())
+	if !ok {
+		return fmt.Errorf("unable to get db data for database %s", ctx.GetCurrentDatabase())
+	}
+
+	rebaseBranch := rebaseBranchWorkingSet.RebaseState().Branch()
+	rebaseWorkingBranch := "dolt_rebase_" + rebaseBranch
+
+	// Check that the branch being rebased hasn't been updated since the rebase started
+	err = validateRebaseBranchHasntChanged(ctx, rebaseBranch, rebaseBranchWorkingSet.RebaseState())
+	if err != nil {
+		return err
+	}
+
+	err = copyABranch(ctx, dbData, rebaseWorkingBranch, rebaseBranch, true, nil)
```

Comment:
You didn't create this problem, but this is a race because the underlying operation (doltdb.NewBranchAtCommit) is not atomic. It sets the branch head, then updates the working set head with whatever it finds there, in two separate optimistic locking noms transactions. So concurrent users of e.g. `call dolt_branch(-f)` and this procedure could overwrite each other's change as well as end up with an inconsistent working set (doesn't match the branch head). 

Probably worth adding a TODO that we need to clean that up at some point, the entire model needs to be overhauled to update more than a single head atomically. We could do part of this now, by changing the NewBranchAtCommit call to use `database.CommitWithWorkingSet` instead of two separate atomic operations.

Reply:
Thanks for pointing this out. I've added a TODO and dug into the code to see where the race condition is. It looks like `database.CommitWithWorkingSet` creates a new commit though, which seems different from what we want here. I'm happy to follow up in another PR and dig in deeper and see if we can get this spot fixed.